kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/KCBQThreadPoolExecutor.java

```diff
@@ -0,0 +1,218 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+
+import com.wepay.kafka.connect.bigquery.config.BigQuerySinkTaskConfig;
+import com.wepay.kafka.connect.bigquery.exception.BigQueryConnectException;
+import io.netty.util.internal.ConcurrentSet;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.connect.sink.SinkTaskContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * ThreadPoolExecutor for writing Rows to BigQuery.
+ *
+ * <p>Keeps track of the number of threads actively writing for each topic.
+ * Keeps track of the number of failed threads in each batch of requests.
+ */
+public class KCBQThreadPoolExecutor extends ThreadPoolExecutor {
+
+  private static final Logger logger = LoggerFactory.getLogger(KCBQThreadPoolExecutor.class);
+
+  private ConcurrentHashMap<String, AtomicInteger> activeThreadTopicCount =
+      new ConcurrentHashMap<>();
+  private ConcurrentSet<Throwable> encounteredErrors = new ConcurrentSet<>();
+  private TopicPartitionManager topicPartitionManager;
+  private int topicThreadLimit;
+
+  /**
+   * @param config the {@link BigQuerySinkTaskConfig}
+   * @param context the {@link SinkTaskContext}
+   * @param workQueue the queue for storing tasks.
+   */
+  public KCBQThreadPoolExecutor(BigQuerySinkTaskConfig config,
+                                SinkTaskContext context,
+                                BlockingQueue<Runnable> workQueue) {
+    super(config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          config.getInt(BigQuerySinkTaskConfig.THREAD_POOL_SIZE_CONFIG),
+          // the following line is irrelevant because the core and max thread counts are the same.
+          1, TimeUnit.SECONDS,
+          workQueue);
+    topicPartitionManager = new TopicPartitionManager(context);
+    topicThreadLimit = config.getInt(BigQuerySinkTaskConfig.TOPIC_MAX_THREADS_CONFIG);
+  }
+
+  @Override
+  protected void beforeExecute(Thread thread, Runnable runnable) {
+    super.beforeExecute(thread, runnable);
+
+    if (runnable instanceof TableWriter) {
+      TableWriter tableWriter = (TableWriter) runnable;
+      String topic = tableWriter.getTopic();
+      activeThreadTopicCount.putIfAbsent(topic, new AtomicInteger(0));
+
+      int topicThreadCount = activeThreadTopicCount.get(topic).incrementAndGet();
+      if (topicThreadCount > topicThreadLimit) {
```

Comment:
Is there an off by one here? At this point, if topic thread limit is 3, and the incrementAndGet returns 3, I don't think that this class will pause the topic. If it doesn't, then the next time around incrementAndGet can return 4, which would mean we'd be running one extra thread.